### PR TITLE
Kontena Storage addon

### PR DESCRIPTION
**Pull request by jakolehm**
Based on https://rook.io/

First release aims for operator & block storage deployments. Other custom resources, like object storage, are supported via plain k8s manifests.

Todo:
- [x] operator
- [x] cluster
- [x] pool
- [x] storageclass
- [x] addon configuration
- [ ] proper cleanup when addon is disabled

## Usage

```yaml
addons:
  kontena-storage:
    enabled: true
    data_dir: /var/lib/kontena-storage
    storage:
      use_all_nodes: true
```

### CODE REVIEW

File: pharos-cluster.gemspec [hunk 1]
```diff
@@ -33,6 +33,7 @@ Gem::Specification.new do |spec|
   spec.add_runtime_dependency "rouge", "~> 3.1"
   spec.add_runtime_dependency "tty-prompt", "~> 0.16"
   spec.add_runtime_dependency "k8s-client", "~> 0.3.2"
+  spec.add_runtime_dependency "awrence", "~> 1.0.0"
```
Comment (jnummelin) [inline pharos-cluster.gemspec]: Where/how is this used? I fail to find any usage of it...

File: non-oss/addons/kontena-storage/addon.rb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          nodes: config.storage&.nodes&.map { |n| n.to_h.to_camelback_keys }
+        },
+        placement: (config.placement || {}).to_h.to_camelback_keys,
+        resources: (config.resources || {}).to_h.to_camelback_keys,
+        dashboard: config.dashboard || { enabled: false }
+      }
+    )
+  end
+
+  def stringify_hash(hash)
```
[the remark this answers is not in the document]
Comment (kke) [reply non-oss/addons/kontena-storage/addon.rb]: Maybe `apply_resources` could stringify or K8s::Resource.to_h could accept something like `to_h(stringify_keys: true)`.

[the remark this answers is not in the document]
Comment (kke) [reply non-oss/addons/kontena-storage/addon.rb]: and `JSON(hash.to_json)` is shorter :)

[the remark this answers is not in the document]
Comment (jakolehm) [reply non-oss/addons/kontena-storage/addon.rb]: Maybe separate pr? I can then merge that implementation here.

[the remark this answers is not in the document]
Comment (kke) [reply non-oss/addons/kontena-storage/addon.rb]: #635

[the remark this answers is not in the document]
Comment (jakolehm) [reply non-oss/addons/kontena-storage/addon.rb]: Merged & removed awrence.

**Comment by jakolehm** [comment]
@jnummelin @kke PTAL

File: Gemfile.lock [hunk 3]
```diff
@@ -19,6 +19,7 @@ GEM
   remote: https://rubygems.org/
   specs:
     ast (2.4.0)
+    awrence (1.0.0)
```
Comment (jnummelin) [inline Gemfile.lock]: No need for this?

Comment (jakolehm) [reply Gemfile.lock]: Fixed.

**Pull request merged**